File: modality-network-consensus/src/sequencing/static_authority.rs
use anyhow::{Result};

use crate::consensus_math::calculate_2f_plus_1;
use crate::election::Election;

use super::Sequencing;

#[derive(Clone)]
pub struct StaticAuthority {
    #[allow(dead_code)]
    election: Election,
    scribes: Vec<String>,  // Assuming scribes are identified by strings - adjust type as needed
}

impl StaticAuthority {
    pub async fn create(scribes: Vec<String>, election: Election) -> Self {
        StaticAuthority {
            election,
            scribes,
        }
    }
}

#[async_trait::async_trait]
impl Sequencing for StaticAuthority {
    async fn get_scribes_at_round(&self, _round: u64) -> Result<Vec<String>> {
        Ok(self.scribes.clone())
    }

    async fn consensus_threshold_for_round(&self, _round: u64) -> Result<u64> {
        Ok(calculate_2f_plus_1(self.scribes.len() as f64))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_static_authority() -> anyhow::Result<()> {
        let scribes = vec!["scribe1".to_string(), "scribe2".to_string(), "scribe3".to_string()];
        let election = Election::RoundRobin(crate::election::round_robin::RoundRobin::create());
        
        let sa = StaticAuthority::create(scribes.clone(), election).await;
        
        let round_scribes = sa.get_scribes_at_round(1).await?;
        assert_eq!(round_scribes, scribes);
        
        let threshold = sa.consensus_threshold_for_round(1).await?;
        assert_eq!(threshold, 3);  // 2f+1 where n=3

        Ok(())
    }
}

File: js/packages/network-consensus/src/sequencing/StaticAuthority.js
import ConsensusMath from "../lib/ConsensusMath.js";
import BlockHeader from "@modality-dev/network-datastore/data/BlockHeader";
export default class StaticAuthority {
  static async create({ datastore, scribes, election }) {
    const sa = new StaticAuthority();
    sa.datastore = datastore;
    sa.election = election;
    sa.scribes = scribes;
    return sa;
  }

  async getScribesAtRound(round) {
    if (round < 0) {
      return [];
    }
    if (!this.scribes) {
      const block_headers = await BlockHeader.findAllInRound({
        datastore: this.datastore,
        round_id: 0,
      });
      const peer_ids = block_headers.map((i) => i.peer_id);
      this.scribes = peer_ids;
    }
    return this.scribes;
  }

  async consensusThresholdForRound(round) {
    if (round < 0) {
      return 0;
    }
    return ConsensusMath.calculate2fplus1(this.scribes.length);
  }
}
